File: CostAnalysisPlanChooser.cs
using System;
using System.Data;
using System.IO;
using System.Windows.Forms;
using Infragistics.Win.UltraWinGrid;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace BossAdmin
{

    public partial class CostAnalysisPlanChooser
    {
        public string msPlanProjectID;
        private Infragistics.Win.UltraWinEditors.UltraOptionSet msOptions = new Infragistics.Win.UltraWinEditors.UltraOptionSet();

        public CostAnalysisPlanChooser()
        {
            InitializeComponent();
        }
        // Dim mdsData As New DataSet()
        // Dim msPathCost As String
        private DataSet GenerateEmptyPlanDataSet()
        {
            var dsData = new DataSet();
            var dtData = new DataTable();

            dtData.Columns.Add(new DataColumn("PlanID", typeof(string)));
            dtData.Columns.Add(new DataColumn("FileName", typeof(string)));
            dtData.Columns.Add(new DataColumn("PlanEstimate", typeof(bool)));
            dtData.Columns.Add(new DataColumn("Source"));
            dtData.Columns.Add(new DataColumn("Use"));
            dsData.Tables.Add(dtData);

            dtData=null;

            return dsData;

        }
        private bool IsInEstimateTable(string sPlanID)
        {
            bool IsInEstimateTableRet = default;
            var rs = new DataSet();
            var dbCalls = new DBCalls();
            try
            {
                IsInEstimateTableRet=dbCalls.GetDataFromSP(ref rs, "spPlanInEstimateTable", sPlanID);
                if (rs.Tables.Count>0)
                {
                    if (rs.Tables[0].Rows.Count>0)
                    {
                        IsInEstimateTableRet=Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(rs.Tables[0].Rows[0][0], 1, false));
                    }
                    else
                    {
                        Interaction.MsgBox("No Records in Plan Estimate table for this plan", MsgBoxStyle.Information, "IsInEstimateTable");
                        IsInEstimateTableRet=false;
                    }
                }
                else
                {
                    IsInEstimateTableRet=false;
                    Interaction.MsgBox(modGlobals.gCN.ConnectionString, MsgBoxStyle.Critical, "IsInEstimateTable");
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox("Error checking for estimate "+ex.Message, MsgBoxStyle.Critical, "IsInEstimateTable");
                IsInEstimateTableRet=false;
            }
            finally
            {
                rs=null;
                dbCalls=null;
            }

            return IsInEstimateTableRet;
        }
        private void CostAnalysisPlanChooser_FormClosing(object sender, FormClosingEventArgs e)
        {
            UltraGrid1.DisplayLayout.Bands[0].SortedColumns.Add("PlanID", false);
        }

        private void CostAnalysisPlanChooser_Load(object sender, EventArgs e)
        {
            if (modGlobals.gsCostAnalysisData is null)
            {
                modGlobals.gsCostAnalysisData=new DataSet();
                modGlobals.gsCostAnalysisData=GenerateEmptyPlanDataSet();
            }
            UltraGrid1.DataSource=modGlobals.gsCostAnalysisData;
            UltraGrid1.DataBind();
            UltraGrid1.DisplayLayout.Bands[0].Columns["FileName"].Width=300;
            UltraGrid1.DisplayLayout.Bands[0].Columns["Planid"].Width=60;
            {
                var withBlock = UltraGrid1.DisplayLayout.Bands[0];
                withBlock.Columns["PlanEstimate"].Width=60;
                withBlock.Columns["PlanEstimate"].Header.Caption="Plan Estimate";
                withBlock.Columns["FileName"].Header.Caption="File Name";
                withBlock.Columns["PlanID"].Header.Caption="Plan";
                withBlock.Columns["Use"].Width=150;
                withBlock.Columns["Source"].Width=100;
                withBlock.Override.WrapHeaderText=Infragistics.Win.DefaultableBoolean.True;
            }
            var valueListItem1 = new Infragistics.Win.ValueListItem();
            var valueListItem2 = new Infragistics.Win.ValueListItem();
            var valueListItem3 = new Infragistics.Win.ValueListItem();

            // Assign a data value for each option
            valueListItem1.DataValue="Plan Estimate";
            valueListItem2.DataValue="File";
            valueListItem3.DataValue="Do Not Use";

            // Assign the text which will be displayed within the UltraGrid for each option
            valueListItem1.DisplayText="Plan Estimate";
            valueListItem2.DisplayText="File";
            valueListItem3.DisplayText="Do Not Use";

            // Add each ValueListItem object to the items collection of the ultraOptionSet
            msOptions.Items.AddRange(new Infragistics.Win.ValueListItem[] { valueListItem1, valueListItem2, valueListItem3 });

            // Set the UltraOptionSet instance as the EditorComponent for the first column of the UltraGrid
            UltraGrid1.DisplayLayout.Bands[0].Columns["Use"].EditorComponent=msOptions;
            modCommon.GetCostPlanPath();

        }

        private void btnProjectPlans_Click(object sender, EventArgs e)
        {
            var dsData = new DataSet();
            // Dim vInPlanEstimate As VariantType
            var dbCalls = new DBCalls();
            string sPlanId;
            string sPath;
            bool bInEstimate;
            bool bInFolder = false;
            try
            {
                Enabled=false;
                My.MyProject.Forms.mdiMain.UseWaitCursor=true;
                if (dbCalls.GetDataFromSP(ref dsData, "spGetPlansForProject", msPlanProjectID, 1))
                {
                    // populate grid
                    foreach (DataRow dRow in dsData.Tables[0].Rows)
                    {
                        // check for files and plan estimate rows
                        sPlanId=(string)(dRow["PlanId"]);
                        // If sPlanId = "2554ab" Then
                        // sPlanId = "2554ab"
                        // End If
                        // If vInPlanEstimate = Nothing Then
                        // bInEstimate = False
                        // Else
                        // bInEstimate = True
                        // End If
                        // 12/2/14 mrb
                        bInEstimate=IsInEstimateTable(sPlanId);
                        sPath=modGlobals.gsPathCost+@"\"+sPlanId+".xlsx";
                        if (modCommon.FileExists(sPath))
                        {
                            bInFolder=true;
                        }
                        else
                        {
                            bInFolder=false;
                            sPath="";
                        }
                        {
                            var withBlock = UltraGrid1;
                            if (bInEstimate==false)
                            {
                                if (bInFolder)
                                {
                                    modGlobals.gsCostAnalysisData.Tables[0].Rows.Add(new object[] { sPlanId, sPath, bInEstimate, "Project Plans", "File" });
                                }
                                else
                                {
                                    modGlobals.gsCostAnalysisData.Tables[0].Rows.Add(new object[] { sPlanId, sPath, bInEstimate, "Project Plans" });
                                }
                            }
                            else
                            {
                                modGlobals.gsCostAnalysisData.Tables[0].Rows.Add(new object[] { sPlanId, sPath, CheckState.Checked, "Project Plans", "Plan Estimate" });
                            }

                        }
                    }
                    UltraGrid1.DisplayLayout.Bands[0].SortedColumns.Add("PlanID", false);
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox("Error in Loading Project Plans: "+ex.Message, MsgBoxStyle.Critical, "btnProjectPlans_Click");
            }
            finally
            {
                Enabled=true;
                My.MyProject.Forms.mdiMain.UseWaitCursor=true;
            }
        }

        private void btnPlansFromFolder_Click(object sender, EventArgs e)
        {
            // Dim vInPlanEstimate As VariantType
            bool bInEstimate;
            string sPlanID;
            string sPath;
            var dbCalls = new DBCalls();
            try
            {
                // GetCostPlanPath()
                Enabled=false;
                My.MyProject.Forms.mdiMain.UseWaitCursor=true;
                var diSource = new DirectoryInfo(modGlobals.gsPathCost);
                if (Directory.Exists(modGlobals.gsPathCost)==true)
                {
                    // Copy each file into it's new directory. 
                    foreach (FileInfo fi in diSource.GetFiles())
                    {
                        if (fi.Extension==".xlsx")
                        {
                            sPlanID=fi.Name.Replace(".xlsx", "");
                            // vInPlanEstimate = dbCalls.GetReturnValueFromSP("spPlanInEstimateTable", sPlanID)
                            bInEstimate=IsInEstimateTable(sPlanID); // 12/2/14

                            sPath=fi.FullName;
                            // If vInPlanEstimate = Nothing Then
                            // bInEstimate = False
                            // Else
                            // bInEstimate = True
                            // End If
                            {
                                var withBlock = UltraGrid1;
                                if (bInEstimate==false)
                                {
                                    modGlobals.gsCostAnalysisData.Tables[0].Rows.Add(new object[] { sPlanID, sPath, bInEstimate, "CostPlan Folder", "File" });
                                }
                                else
                                {
                                    modGlobals.gsCostAnalysisData.Tables[0].Rows.Add(new object[] { sPlanID, sPath, CheckState.Checked, "CostPlan Folder", "Plan Estimate" });
                                }
                            }
                        }
                    }
                    UltraGrid1.DisplayLayout.Bands[0].SortedColumns.Add("PlanID", false);
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox("Error listing plan files in folder: "+ex.Message, MsgBoxStyle.Critical, "btnPlansFromFolder_Click");
            }
            finally
            {
                Enabled=true;
                My.MyProject.Forms.mdiMain.UseWaitCursor=false;
            }
        }

        private void UltraGrid1_AfterCellUpdate(object sender, CellEventArgs e)
        {
            if (e.Cell.Column.Key=="PlanID"&!string.IsNullOrEmpty(e.Cell.Text))
            {
                if (IsInEstimateTable(e.Cell.Text))
                {
                    UltraGrid1.DisplayLayout.ActiveRow.Cells["PlanEstimate"].Value=CheckState.Checked;
                    UltraGrid1.DisplayLayout.ActiveRow.Cells["Use"].Value="Plan Estimate";
                }
                else
                {
                    UltraGrid1.DisplayLayout.ActiveRow.Cells["PlanEstimate"].Value=CheckState.Unchecked;
                }
            }
        }

        private void UltraGrid1_AfterRowInsert(object sender, RowEventArgs e)
        {
            UltraGrid1.DisplayLayout.ActiveRow.Cells["Source"].Value="User Added";

        }


        private void btnSavePlans_Click(object sender, EventArgs e)
        {
            Close();
        }

        private void btnClearAllPlans_Click(object sender, EventArgs e)
        {
            modGlobals.gsCostAnalysisData=null;
            modGlobals.gsCostAnalysisData=GenerateEmptyPlanDataSet();
            UltraGrid1.DataSource=modGlobals.gsCostAnalysisData;
            UltraGrid1.DataBind();

        }
    }
}

File: CostAnalysisPlanChooser.vb
Option Explicit On
Imports BossAdmin.DBCalls
Imports System.Data
Imports System
Imports Infragistics.Win.UltraWinGrid
Imports System.IO

Public Class CostAnalysisPlanChooser
    Public msPlanProjectID As String
    Dim msOptions As New Infragistics.Win.UltraWinEditors.UltraOptionSet
    ' Dim mdsData As New DataSet()
    'Dim msPathCost As String
    Private Function GenerateEmptyPlanDataSet() As DataSet
        Dim dsData As New DataSet()
        Dim dtData As New DataTable()

        dtData.Columns.Add(New DataColumn("PlanID", GetType(String)))
        dtData.Columns.Add(New DataColumn("FileName", GetType(String)))
        dtData.Columns.Add(New DataColumn("PlanEstimate", GetType(Boolean)))
        dtData.Columns.Add(New DataColumn("Source"))
        dtData.Columns.Add(New DataColumn("Use"))
        dsData.Tables.Add(dtData)

        dtData = Nothing

        Return dsData

    End Function
    Private Function IsInEstimateTable(ByVal sPlanID As String) As Boolean
        Dim rs As New DataSet
        Dim dbCalls As New DBCalls
        Try
            IsInEstimateTable = dbCalls.GetDataFromSP(rs, "spPlanInEstimateTable", sPlanID)
            If Rs.tables.count > 0 Then
                If rs.Tables(0).Rows.Count > 0 Then
                    IsInEstimateTable = rs.Tables(0).Rows(0).Item(0) = 1
                Else
                    MsgBox("No Records in Plan Estimate table for this plan", MsgBoxStyle.Information, "IsInEstimateTable")
                    IsInEstimateTable = False
                End If
            Else
                IsInEstimateTable = False
                MsgBox(gCN.ConnectionString, MsgBoxStyle.Critical, "IsInEstimateTable")
            End If
        Catch ex As Exception
            MsgBox("Error checking for estimate " & ex.Message, MsgBoxStyle.Critical, "IsInEstimateTable")
            IsInEstimateTable = False
        Finally
            rs = Nothing
            dbCalls = Nothing
        End Try
    End Function
    Private Sub CostAnalysisPlanChooser_FormClosing(sender As Object, e As FormClosingEventArgs) Handles Me.FormClosing
        UltraGrid1.DisplayLayout.Bands(0).SortedColumns.Add("PlanID", False)
    End Sub

    Private Sub CostAnalysisPlanChooser_Load(sender As Object, e As EventArgs) Handles Me.Load
        If gsCostAnalysisData Is Nothing Then
            gsCostAnalysisData = New DataSet
            gsCostAnalysisData = GenerateEmptyPlanDataSet()
        End If
        UltraGrid1.DataSource = gsCostAnalysisData
        UltraGrid1.DataBind()
        UltraGrid1.DisplayLayout.Bands(0).Columns("FileName").Width = 300
        UltraGrid1.DisplayLayout.Bands(0).Columns("Planid").Width = 60
        With UltraGrid1.DisplayLayout.Bands(0)
            .Columns("PlanEstimate").Width = 60
            .Columns("PlanEstimate").Header.Caption = "Plan Estimate"
            .Columns("FileName").Header.Caption = "File Name"
            .Columns("PlanID").Header.Caption = "Plan"
            .Columns("Use").Width = 150
            .Columns("Source").Width = 100
            .Override.WrapHeaderText = Infragistics.Win.DefaultableBoolean.True
        End With
        Dim valueListItem1 As New Infragistics.Win.ValueListItem()
        Dim valueListItem2 As New Infragistics.Win.ValueListItem()
        Dim valueListItem3 As New Infragistics.Win.ValueListItem()

        'Assign a data value for each option
        valueListItem1.DataValue = "Plan Estimate"
        valueListItem2.DataValue = "File"
        valueListItem3.DataValue = "Do Not Use"

        'Assign the text which will be displayed within the UltraGrid for each option
        valueListItem1.DisplayText = "Plan Estimate"
        valueListItem2.DisplayText = "File"
        valueListItem3.DisplayText = "Do Not Use"

        'Add each ValueListItem object to the items collection of the ultraOptionSet
        msOptions.Items.AddRange(New Infragistics.Win.ValueListItem() {valueListItem1, valueListItem2, valueListItem3})

        'Set the UltraOptionSet instance as the EditorComponent for the first column of the UltraGrid
        UltraGrid1.DisplayLayout.Bands(0).Columns("Use").EditorComponent = msOptions
        GetCostPlanPath()

    End Sub

    Private Sub btnProjectPlans_Click(sender As Object, e As EventArgs) Handles btnProjectPlans.Click
        Dim dsData As New DataSet
        Dim dbCalls As New DBCalls
        '  Dim vInPlanEstimate As VariantType
        Dim dRow As DataRow
        Dim sPlanId As String
        Dim sPath As String
        Dim bInEstimate As Boolean
        Dim bInFolder As Boolean = False
        Try
            Me.Enabled = False
            mdiMain.UseWaitCursor = True
            If dbCalls.GetDataFromSP(dsData, "spGetPlansForProject", msPlanProjectID, 1) Then
                'populate grid
                For Each dRow In dsData.Tables(0).Rows
                    'check for files and plan estimate rows
                    sPlanId = Trim(dRow.Item("PlanId"))
                    'If sPlanId = "2554ab" Then
                    '    sPlanId = "2554ab"
                    'End If
                    'If vInPlanEstimate = Nothing Then
                    '    bInEstimate = False
                    'Else
                    '    bInEstimate = True
                    'End If
                    '12/2/14 mrb
                    bInEstimate = IsInEstimateTable(sPlanId)
                    sPath = gsPathCost & "\" & sPlanId & ".xlsx"
                    If FileExists(sPath) Then
                        bInFolder = True
                    Else
                        bInFolder = False
                        sPath = ""
                    End If
            With UltraGrid1
                        If bInEstimate = False Then
                            If bInFolder Then
                                gsCostAnalysisData.Tables(0).Rows.Add(New Object() {sPlanId, sPath, bInEstimate, "Project Plans", "File"})
                            Else
                                gsCostAnalysisData.Tables(0).Rows.Add(New Object() {sPlanId, sPath, bInEstimate, "Project Plans"})
                            End If
                        Else
                            gsCostAnalysisData.Tables(0).Rows.Add(New Object() {sPlanId, sPath, CheckState.Checked, "Project Plans", "Plan Estimate"})
                        End If

            End With
                Next
                UltraGrid1.DisplayLayout.Bands(0).SortedColumns.Add("PlanID", False)
            End If
        Catch ex As Exception
            MsgBox("Error in Loading Project Plans: " & ex.Message, MsgBoxStyle.Critical, "btnProjectPlans_Click")
        Finally
            Me.Enabled = True
            mdiMain.UseWaitCursor = True
        End Try
    End Sub

    Private Sub btnPlansFromFolder_Click(sender As Object, e As EventArgs) Handles btnPlansFromFolder.Click
        'Dim vInPlanEstimate As VariantType
        Dim bInEstimate As Boolean
        Dim sPlanID As String
        Dim sPath As String
        Dim dbCalls As New DBCalls
        Try
            ' GetCostPlanPath()
            Me.Enabled = False
            mdiMain.UseWaitCursor = True
            Dim diSource As DirectoryInfo = New DirectoryInfo(gsPathCost)
            If Directory.Exists(gsPathCost) = True Then
                ' Copy each file into it's new directory. 
                For Each fi As FileInfo In diSource.GetFiles()
                    If fi.Extension = ".xlsx" Then
                        sPlanID = fi.Name.Replace(".xlsx", "")
                        'vInPlanEstimate = dbCalls.GetReturnValueFromSP("spPlanInEstimateTable", sPlanID)
                        bInEstimate = IsInEstimateTable(sPlanID) '12/2/14

                        sPath = fi.FullName
                        'If vInPlanEstimate = Nothing Then
                        '    bInEstimate = False
                        'Else
                        '    bInEstimate = True
                        'End If
                        With UltraGrid1
                            If bInEstimate = False Then
                                gsCostAnalysisData.Tables(0).Rows.Add(New Object() {sPlanID, sPath, bInEstimate, "CostPlan Folder", "File"})
                            Else
                                gsCostAnalysisData.Tables(0).Rows.Add(New Object() {sPlanID, sPath, CheckState.Checked, "CostPlan Folder", "Plan Estimate"})
                            End If
                        End With
                    End If
                Next
                UltraGrid1.DisplayLayout.Bands(0).SortedColumns.Add("PlanID", False)
            End If
        Catch ex As Exception
            MsgBox("Error listing plan files in folder: " & ex.Message, MsgBoxStyle.Critical, "btnPlansFromFolder_Click")
        Finally
            Me.Enabled = True
            mdiMain.UseWaitCursor = False
        End Try
    End Sub

    Private Sub UltraGrid1_AfterCellUpdate(sender As Object, e As CellEventArgs) Handles UltraGrid1.AfterCellUpdate
        If e.Cell.Column.Key = "PlanID" And e.Cell.Text <> "" Then
            If IsInEstimateTable(e.Cell.Text) Then
                UltraGrid1.DisplayLayout.ActiveRow.Cells("PlanEstimate").Value = CheckState.Checked
                UltraGrid1.DisplayLayout.ActiveRow.Cells("Use").Value = "Plan Estimate"
            Else
                UltraGrid1.DisplayLayout.ActiveRow.Cells("PlanEstimate").Value = CheckState.Unchecked
            End If
        End If
    End Sub

    Private Sub UltraGrid1_AfterRowInsert(sender As Object, e As RowEventArgs) Handles UltraGrid1.AfterRowInsert
        UltraGrid1.DisplayLayout.ActiveRow.Cells("Source").Value = "User Added"
 
    End Sub

 
    Private Sub btnSavePlans_Click(sender As Object, e As EventArgs) Handles btnSavePlans.Click
        Me.Close()
    End Sub

    Private Sub btnClearAllPlans_Click(sender As Object, e As EventArgs) Handles btnClearAllPlans.Click
        gsCostAnalysisData = Nothing
        gsCostAnalysisData = GenerateEmptyPlanDataSet()
        UltraGrid1.DataSource = gsCostAnalysisData
        UltraGrid1.DataBind()

    End Sub
End Class